utils/yoke/src/cart.rs

```diff
@@ -0,0 +1,117 @@
+///
+/// The essential invariant that must be maintained by implementors is that `Self::Inner` references
+/// obtained from this type via `Self::get_inner()` must continue to be valid for the
+/// duration of this type provided that this type is never accessed via `&mut` references.
+///
+/// For example, `Vec<u8>` is a valid [`Cart`], however `Vec<RefCell<u8>>` is not, because
+/// in the latter type it is possible to use interior mutation to change the data.
+///
+/// In general, this means "no interior mutability", though interior mutability can be fine
+/// if the interior mutability is to something that does not affect references. For example,
+/// `Rc<[u8]>` does have interior mutability in the refcount, but you cannot get references
+/// to it, so it's fine. On the other hand, `Weak<Box<[u8]>>` cannot be a valid [`Cart`] because
+/// it is possible for the backing buffer to be cleaned up without warning.
+///
+/// Common [`Cart`] types to use with [`Yoke`] are ones wrapping `[u8]` or `str`
+/// (`Box<[u8]>`, `Rc<[u8]>`, etc) since those are typical inputs to zero-copy
+/// deserialization and parsing.
+///
+/// Typically the [`Cart`] trait will be implemented by smart pointers whilst [`Cartable`] is used
+/// as a helper to talk about types that have the same property provided they are not moved, sucn that
+/// [`Cart`] can be implemented on wrappers that can maintain this property even if they are moved.
+pub unsafe trait Cart {
+    type Inner: ?Sized;
+    /// Get the inner data
+    fn get_inner(&self) -> &Self::Inner;
+}
+
+/// [`Cartable`] is a helper trait for implementng [`Cart`]. It has a similar invariant:
+/// all references obtained from this type must continue to be valid for the lifetime
+/// of this type provided that this type is never moved or accessed via an `&mut` reference.
+///
+/// Essentially, this means "no interior mutability", however interior mutability which cannot be
+/// seen from the outside is fine.
+pub unsafe trait Cartable {}
+
+unsafe impl<T: Cartable + ?Sized> Cart for Rc<T> {
+    type Inner = T;
+    fn get_inner(&self) -> &Self::Inner {
+        &**self
+    }
+}
+
+unsafe impl<T: Cartable + ?Sized> Cart for Arc<T> {
+    type Inner = T;
+    fn get_inner(&self) -> &Self::Inner {
+        &**self
+    }
+}
+
+unsafe impl<T: Cartable + ?Sized> Cart for Box<T> {
+    type Inner = T;
+    fn get_inner(&self) -> &Self::Inner {
+        &**self
+    }
+}
+
+unsafe impl<T: Cartable> Cart for Vec<T> {
+    type Inner = [T];
+    fn get_inner(&self) -> &Self::Inner {
+        &**self
+    }
+}
+
+unsafe impl Cart for String {
+    type Inner = str;
+    fn get_inner(&self) -> &Self::Inner {
+        &**self
+    }
+}
+
+// useful for constructing `Yokeable`s that may be borrowed
+// or owned, this is basically intended to be used with `Yoke::new()`
+unsafe impl<T: Cart> Cart for Option<T> {
+    type Inner = T::Inner;
+    fn get_inner(&self) -> &Self::Inner {
+        self.as_ref().unwrap().get_inner()
```

Comment:
Unsure! I don't like this either, we could perhaps instead do this:

 - `Yoke` no longer requires `C: Cart`
 - `attach_to_cart()` does instead
 - `attach_to_cart()` has an `attach_to_maybe_cart()` version that stuffs it in an `Option`
 - `Yokeable::new` is also only implemented for the `Option` case.


I did it the current way because I figured that others may want to do similar tricks for other cart types. But perhaps `Option` is the only cart wrapper that makes sense!

Reply:
I did this